MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs

```diff
@@ -2779,6 +2878,73 @@ private string GetFixedSizeScalingFilter(Video3DFormat? threedFormat, int reques
                         request.MaxHeight));
             }
 
+            // Add Cuda tonemapping filter.
+            if (isTonemappingSupportedOnNvenc && !isSwDecoder && isCudaTonemappingSupported)
+            {
+                if (string.Equals(videoStream.ColorTransfer, "arib-std-b67", StringComparison.OrdinalIgnoreCase))
+                {
+                    // HLG
+                    filters.Add("setparams=color_primaries=bt2020:color_trc=arib-std-b67:colorspace=bt2020nc");
+                }
+                else
+                {
+                    // HDR10
+                    filters.Add("setparams=color_primaries=bt2020:color_trc=smpte2084:colorspace=bt2020nc");
+                }
+
+                var parameters = (hasGraphicalSubs && isCudaOverlaySupported && isNvencEncoder)
+                    ? "tonemap_cuda=format=yuv420p:primaries=bt709:transfer=bt709:matrix=bt709:tonemap={0}:peak={1}:desat={2}"
+                    : "tonemap_cuda=format=nv12:primaries=bt709:transfer=bt709:matrix=bt709:tonemap={0}:peak={1}:desat={2}";
+
+                if (options.TonemappingParam != 0)
+                {
+                    parameters += ":param={3}";
+                }
+
+                if (!string.Equals(options.TonemappingRange, "auto", StringComparison.OrdinalIgnoreCase))
+                {
+                    parameters += ":range={4}";
+                }
+
+                filters.Add(
+                    string.Format(
+                        CultureInfo.InvariantCulture,
+                        parameters,
+                        options.TonemappingAlgorithm,
+                        options.TonemappingPeak,
+                        options.TonemappingDesat,
+                        options.TonemappingParam,
+                        options.TonemappingRange));
+
+                if (isLibX264Encoder
+                    || isLibX265Encoder
+                    || hasTextSubs
+                    || (hasGraphicalSubs && !isCudaOverlaySupported && isNvencEncoder))
+                {
+                    if (isNvencEncoder)
+                    {
+                        isHwuploadCudaRequired = true;
+                    }
+
+                    filters.Add("hwdownload");
+                    filters.Add("format=nv12");
+                }
+
+                // SDR
+                parameters = "setparams=color_primaries=bt709:color_trc=bt709:colorspace=bt709";
```

Comment:
Duplicated code? It's also at line 2756. Could be put in a separate static method.